### PR TITLE
Add Devise

### PR DESCRIPTION
**Pull request by Carshy**
In the current project, I added the devise gem to my app and use it for the registration and login of users.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,4 +1,8 @@
 class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
```
Comment (tufoinnkuo10) [inline app/models/user.rb]: - [ ] Great job so far. Please ensure that the application asks for confirmation of email.
 `:confirmable` is missing in your user model. Please fix that. it helps sends emails with confirmation instructions and verifies whether an account is already confirmed during sign in

**Review by tufoinnkuo10 — CHANGES_REQUESTED** [review]
Hi @Carshy 

There are some issues that you still need to work on to go to the next project but you are almost there!

To Highlight
✅ You used the correct Github flow
✅ Descriptive pull request message
✅ linter error checks validated
✅ Devise well added

Required Changes ♻️
Check the comments under the review
Optional suggestions
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using this [form.](https://airtable.com/shrQAqnBwek5a0O0s)

**Comment by Carshy** [comment]
Changes made as requested. @tufoinnkuo10 thank you for your review it was really helpful

File: app/models/user.rb [hunk 2]
```diff
@@ -1,4 +1,8 @@
 class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable, :confirmable
```
Comment (codepantha) [inline app/models/user.rb:5]: - [ ] An error now occurs when a user tries to sign up. Please take a look at the error image attached below and try to fix it 👇 
image

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: REQUIRES CHANGES ♻️ 

## Hi @Carshy 👋, 

Great work on making the changes requested by the previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



## Highlights 🎯 
✔️ Used GitFlow
✔️ Good commit history
✔️ No linter errors

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Carshy** [comment]
Changes made as requested by the previous reviewer. Confirmable now working okay. @codepantha thanks for your review it was really helpful

**Comment by codepantha** [comment]
I'm glad I could help. Happy coding.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,5 +1,19 @@
...
+    devise_parameter_sanitizer.permit(:sign_up) { |u| u.permit(:name, :surname, :bio, :photo, :email, :password) }
+    devise_parameter_sanitizer.permit(:account_update) do |u|
+      u.permit(:name, :surname, :bio, :photo, :email, :password, :current_password)
+    end
   end
+
+  # Catch all CanCan errors and alert the user of the exception
+  # rescue_from CanCan::AccessDenied do |exception|
+  #   redirect_to root_url, alert: exception.message
+  # end
```
Comment (rustam-tolipov) [inline app/controllers/application_controller.rb]: - [ ] To make your code professional, kindly remove commented codes from your file :)

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,5 +1,19 @@
 class ApplicationController < ActionController::Base
-  def current_user
-    @current_user = User.first
+  protect_from_forgery with: :exception
+  # before_action :authenticate_user!
```
Comment (rustam-tolipov) [inline app/controllers/application_controller.rb]: - [ ] I don't think your `devise` is necessary if you do not authenticate your user. Without `authenticate_user!` your `devise` won't be able to do its job. For more information: what is [authenticate_user](https://www.google.com/search?q=what+is+authenticate_user+rails&rlz=1C5CHFA_enUZ1015UZ1015&sxsrf=ALiCzsZ8udM_E95KySVvzoqUGjdVaJf53g%3A1668184045400&ei=7XduY-L4F5rQqwH_5a-wBQ&oq=whatauthenticate_user+rails&gs_lcp=Cgxnd3Mtd2l6LXNlcnAQARgAMgoIIRDDBBAKEKABOgoIABBHENYEELADOg0IABDkAhDWBBCwAxgBSgQITRgBSgQIQRgASgQIRhgBUPgHWJARYK8caAFwAXgAgAG-AYgBnAWSAQMwLjSYAQCgAQHIAQ3AAQHaAQYIARABGAk&sclient=gws-wiz-serp#:~:text=Devise%20also%20comes%20with%20some%20very%20useful%20helper%20functions%3A%20before_action%20%3Aauthenticate_user!%20%E2%80%94%20Add%20to%20any%20controller%20to%20limit%20access%20to%20an%20action%20unless%20a%20user%20is%20logged%20in.) ? 📃

Kindly uncomment this part of your code :)

**Review by rustam-tolipov — CHANGES_REQUESTED** [review]
Hello @Carshy, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
_Your project is almost complete!_
_No Linters errors_
_Kudos for adding MIT license to your project_

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
Kindly personalize your MIT license 🪪 👮🏻‍♂️

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Carshy** [comment]
Changes made as requested. @Rustamxon7 thanks for your help very much appreciated

File: db/migrate/20221111152908_add_confirmable_to_devise.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class AddConfirmableToDevise < ActiveRecord::Migration[7.0]
+  def up
+    add_column :users, :confirmation_token, :string
+    add_column :users, :confirmed_at, :datetime
+    add_column :users, :confirmation_sent_at, :datetime
+    add_column :users, :unconfirmed_email, :string # Only if using reconfirmable
```
Comment (Ismailco) [inline db/migrate/20221111152908_add_confirmable_to_devise.rb:6]: - [ ] When creating migrations I got this error about `reconfirmable`, kindly fix this so the migrations are created correctly

Screenshot_20221111_184612

**Review by Ismailco — CHANGES_REQUESTED** [review]
### Hi @Carshy, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights

✔️  No linter errors
✔️ Used Gitflow
✔️ Work documented professionally

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Carshy** [comment]
All changes made as requested. @Ismailco thank you for your review it was really helpful

**Review by afanimartin — APPROVED** [review]
Hi @Carshy, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 🎉 🎉 🎉 
- Changes made as requested

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

### Martin Afani
#### Slack: whiz25

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**